test/lit/passes/vacuum-tnh.wast

```diff
@@ -117,12 +190,41 @@
     )
   )
 
-  ;; CHECK:      (func $toplevel
-  ;; CHECK-NEXT:  (nop)
-  ;; CHECK-NEXT: )
+  ;; YESTNH:      (func $toplevel
+  ;; YESTNH-NEXT:  (nop)
+  ;; YESTNH-NEXT: )
+  ;; NO_TNH:      (func $toplevel
+  ;; NO_TNH-NEXT:  (unreachable)
+  ;; NO_TNH-NEXT: )
   (func $toplevel
     ;; A removable side effect at the top level of a function. We can turn this
     ;; into a nop.
     (unreachable)
   )
+
+  ;; YESTNH:      (func $drop-loop
+  ;; YESTNH-NEXT:  (nop)
+  ;; YESTNH-NEXT: )
+  ;; NO_TNH:      (func $drop-loop
+  ;; NO_TNH-NEXT:  (drop
+  ;; NO_TNH-NEXT:   (loop $loop (result i32)
+  ;; NO_TNH-NEXT:    (br_if $loop
+  ;; NO_TNH-NEXT:     (i32.const 1)
+  ;; NO_TNH-NEXT:    )
+  ;; NO_TNH-NEXT:    (i32.const 10)
+  ;; NO_TNH-NEXT:   )
+  ;; NO_TNH-NEXT:  )
+  ;; NO_TNH-NEXT: )
+  (func $drop-loop
+    ;; A loop has effects, since it might infinite loop (and hit a timeout trap
+    ;; eventually), so we do not vacuum it out unless we are ignoring traps.
+    (drop
+      (loop $loop (result i32)
+        (br_if $loop
+          (i32.const 1)
```

Comment:
What happens if the loop contains atomic operations?

Reply:
Atomics are an additional effect, so it wouldn't optimize out. I added a testcase now.